### PR TITLE
7789 controller fixes take 2

### PR DESCRIPTION
**Pull request by clydebarrow**
# What does this implement/fix?

The 7789 driver was set up to use mode 3 SPI - high clock, trailing edge. The datasheet says it should use mode 0 - low clock, leading edge. Implementing this negates the need for the extra SLPOUT as per the previous PR.

SPI clock has also been bumped to 40 MHz - the datasheet says clock high and low periods are minimum 7ns, which means the maximum clock rate is around 70MHz.

This has been tested against the current dev branch (Arduino only) as well as PR https://github.com/esphome/esphome/pull/5311 (Arduino and ESP-IDF 5.0).

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)

**Related issue or feature (if applicable):** Relates to: https://github.com/esphome/feature-requests/issues/901

## Test Environment

- [x] ESP32
- [x] ESP32 IDF

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
esphome:
  name: t-embed

esp32:
  board: esp32-s3-devkitc-1
  variant: esp32s3
  framework:
    type: arduino

logger:
api:
ota:
  password: !secret ota_password
wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password

power_supply:
  - id: power_on
    pin: GPIO46

spi:
  clk_pin: GPIO12
  mosi_pin: GPIO11

display:
  - platform: st7789v
    power_supply: power_on
    model: "Custom"
    eightbitcolor: False
    rotation: 270
    width: 170
    height: 320
    offset_width: 0
    offset_height: 35
    backlight_pin: GPIO15
    cs_pin: GPIO10
    dc_pin: GPIO13
    reset_pin: GPIO9
    id: disp
    lambda: |-
      auto rand_col = Color::random_color();
      it.filled_rectangle(0, 0, it.get_width() - 1, it.get_height() - 1, rand_col);
```

## Checklist:
  - [x] The code change is tested and works locally.

### CODE REVIEW

**Comment by kbx81** [comment]
The issue with clock rates higher than 20MHz is this:

```
[13:40:36][D][esp-idf:000]: E (331) spi_hal: spi_hal_cal_clock_conf(94): When work in full-duplex mode at frequency > 26.7MHz, device cannot read correct data.
[13:40:36]Try to use IOMUX pins to increase the frequency limit, or use the half duplex mode.
[13:40:36]Please note the SPI master can only work at divisors of 80MHz, and the driver always tries to find the closest frequency to your configuration.
[13:40:36]Specify ``SPI_DEVICE_NO_DUMMY`` to ignore this checking. Then you can output data at higher speed, or read data at your
[13:40:36][D][esp-idf:000]: E (371) spi_master: spi_bus_add_device(366): assigned clock speed not supported
[13:40:36]
[13:40:36][E][spi-esp-idf:042]: Add device failed - err 106
[13:40:36][W][spi-esp-idf:053]: spi_setup called before initialisation
[13:40:36][D][esp-idf:000]: E (422) spi_master: check_trans_valid(684): invalid dev handle
[13:40:36]
[13:40:36][E][spi-esp-idf:085]: Transmit failed - err 102
```

On some boards, the display is hardwired and the pins cannot be changed -- and we don't currently have a way to change the data rate via YAML (although, in one case, this has been worked around -- see the ILI-series component).

Regardless, I think we should stick to a data rate that by default does not render errors. Thoughts? 😄

**Comment by clydebarrow** [comment]
> Regardless, I think we should stick to a data rate that by default does not render errors. Thoughts? 😄

Absolutely agree. With my new SPI code it would be possible to add a YAML config to change it, then have the best of all worlds.

**Comment by clydebarrow** [comment]
I did try to test on other pins, and did not see that message, but the UART logging on my S3 board isn;t working well and logging via the network misses the setup phase. It would explain why I saw no clock though :-o

**Comment by kbx81** [comment]
> I did try to test on other pins, and did not see that message, but the UART logging on my S3 board isn;t working well and logging via the network misses the setup phase. It would explain why I saw no clock though :-o

Indeed, it seems that particular message only appears once at init time and never again...and the result is that the SPI device/client fails to register and so any communication that follows simply fails -- and does so silently if you don't happen to be watching the logs. 😇

File: esphome/components/st7789v/st7789v.cpp [hunk 1]
```diff
@@ -5,6 +5,7 @@ namespace esphome {
 namespace st7789v {
 
 static const char *const TAG = "st7789v";
+static const size_t TEMP_BUFFER_SIZE = 512;
```
Comment (kbx81) [inline esphome/components/st7789v/st7789v.cpp]: Out of concern for memory-constrained devices (specifically the '8266) how do you feel about making this a little smaller? ...maybe 256 or even 128? I realize larger is better for performance, though, so there is a trade-off. Thoughts?

Comment (clydebarrow) [reply esphome/components/st7789v/st7789v.cpp]: I tested initially with 128 and it was pretty good, the gain going to 512 was noticeable but not huge. So 128 it is.

**Review by kbx81 — APPROVED** [review]

File: esphome/components/st7789v/st7789v.cpp (hunk 1 above)
Comment (clydebarrow) [reply esphome/components/st7789v/st7789v.cpp]: At least with memory-constrained devices all the RAM is DMA-capable :-)

Comment (kbx81) [reply esphome/components/st7789v/st7789v.cpp]: Thanks again! 🍻

**Pull request merged**